assets/js/atomic/blocks/product/add-to-cart/product-types/variable/variation-attributes/utils.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * External dependencies
+ */
+import { keyBy } from 'lodash';
+
+/**
+ * Key an array of attributes by name,
+ *
+ * @param {Object} attributes Attributes array.
+ */
+export const getAttributes = ( attributes ) => {
+	return attributes
+		? keyBy(
+				Object.values( attributes ).filter(
+					( { has_variations: hasVariations } ) => hasVariations
+				),
+				'name'
+		  )
+		: [];
+};
+
+/**
+ * Format variations from the API into a map of just the attribute names and values.
+ *
+ * @param {Array} variations Variations array.
+ */
+export const getVariationAttributes = ( variations ) => {
+	if ( ! variations ) {
+		return {};
+	}
+
+	const attributesMap = {};
+
+	variations.forEach( ( { id, attributes } ) => {
+		attributesMap[ id ] = attributes.reduce( ( acc, { name, value } ) => {
+			acc[ name ] = value;
+			return acc;
+		}, [] );
+	} );
+
+	return attributesMap;
+};
+
+/**
+ * Given a list of terms, filter them and return options for the select boxes.
+ *
+ * @param {Object} attributeTerms List of attribute term objects.
+ * @param {?Array} validAttributeTerms Valid values if selections have been made already.
+ * @return {Array} Value/Label pairs of select box options.
+ */
+export const getSelectControlOptions = (
+	attributeTerms,
+	validAttributeTerms = null
+) => {
+	return Object.values( attributeTerms )
+		.map( ( { name, slug } ) => {
+			if (
+				validAttributeTerms === null ||
+				validAttributeTerms.includes( null ) ||
+				validAttributeTerms.includes( slug )
+			) {
+				return {
+					value: slug,
+					label: name,
+				};
+			}
+			return null;
+		} )
+		.filter( Boolean );
+};
+
+/**
+ * Given a list of variations and a list of attribute values, return variations which match.
+ *
+ * Allows an attribute to be excluded by name. This is used to filter displayed options for
+ * individual attribute selects.
+ *
+ * @param {Object} props
+ * @param {Object} props.selectedAttributes List of selected attributes.
+ * @param {Object} props.variationAttributes List of variations and their attributes.
+ * @param {Object} props.attributeNames List of all possible attribute names.
+ * @param {string} [props.exclude] Optionally exclude an attribute from the matching by name.
+ * @return {Array} List of matching variation IDs.
+ */
+export const getVariationsMatchingSelectedAttributes = ( {
+	selectedAttributes,
+	variationAttributes,
+	attributeNames,
+	exclude = '',
+} ) => {
+	return Object.keys( variationAttributes ).filter( ( variationId ) =>
+		attributeNames
+			.filter( ( attributeName ) => attributeName !== exclude )
```

Comment:
Instead of filtering the attribute names here, why not just passing to this function the `attributeNames` array already filtered? This would avoid looping through the array inside another loop (`filter` inside a `filter`) and would simplify the function signature with one less argument.